### PR TITLE
Replace http.error.reason with OTel standard error.type

### LINKED ISSUE #91909: [Http Metrics] Replace http.error.reason with OTel standard error.type
Align attribute name and semantics with the OpenTelemetry spec (which was finalized yesterday 2023/9/11 - see https://github.com/open-telemetry/semantic-conventions/pull/205) -- `http.error.reason` attribute has been standardized under name `error.type`. We need to:
- Rename `http.error.reason` to `error.type`.
- Report `error.type` also for valid responses where the HTTP status code indicates an error (4xx and 5xx) -- in such case, the value should be the string representation of the HTTP status code.
    - Note: Currently we report `http.error.reason` only when the underlying handler fails to fetch a response.

See `error.type` in [the spec](https://github.com/open-telemetry/semantic-conventions/blob/2bad9afad58fbd6b33cc683d1ad1f006e35e4a5d/docs/http/http-spans.md#common-attributes) for more details.

For context: We introduced the attribute `http.error.reason` in PR #89809 (on 2023/8/3) based on the in-progress draft of the OTel spec (https://github.com/open-telemetry/semantic-conventions/pull/205).

Given that this is new 8.0 feature, we should adapt to the spec in 8.0 to avoid breaking changes with 9.0.

### PR DESCRIPTION
**Pull request by antonfirsov**
Fixes #91909

Align attribute name and semantics with the OpenTelemetry spec (which was finalized yesterday 2023/9/11 - see https://github.com/open-telemetry/semantic-conventions/pull/205) -- `http.error.reason` attribute has been standardized under name `error.type`. We need to:
- Rename `http.error.reason` to `error.type`.
- Report `error.type` also for valid responses where the HTTP status code indicates an error (4xx or 5xx) -- in such case, the value should be the string representation of the HTTP status code.
    - Note: Currently we report `http.error.reason` only when the underlying handler fails to fetch a response.

See `error.type` in [the spec](https://github.com/open-telemetry/semantic-conventions/blob/2bad9afad58fbd6b33cc683d1ad1f006e35e4a5d/docs/http/http-spans.md#common-attributes) for more details.

For context: We introduced the attribute `http.error.reason` in PR #89809 (on 2023/8/3) based on the in-progress draft of the OTel spec (https://github.com/open-telemetry/semantic-conventions/pull/205).

Given that this is new 8.0 feature, we should adapt to the spec in 8.0 to avoid breaking changes with 9.0.

### CODE REVIEW

**Comment by danmoseley** [comment]
Is there any similar change needed in Kestrel @JamesNK

**Comment by JamesNK** [comment]
No aspnetcore changes are required. This attribute is for HTTP clients only.

**Comment by danmoseley** [comment]
Ok. Yes sounds like we should definitely get into 8.0 as suggested

**Comment by karelz** [comment]
Let's get this into 9.0 first, then we can immediately create backport to 8.0. IMO it should not have a problem to get in.

@artl93 heads up. If you have concerns, please let us know ASAP.

File: src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs [hunk 1]
```diff
@@ -209,6 +221,14 @@ private static object GetBoxedStatusCode(int statusCode)
                 : statusCode;
         }
 
+        private static string GetStatusCodeString(int statusCode)
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs:224]: @stephentoub should we consider bumping `SmallNumberCacheLength` to 512 instead and we can delete this code?

**Review by MihaZupan — APPROVED** [review]
Looks good, thank you

File: src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs [hunk 2]
```diff
@@ -209,6 +221,14 @@ private static object GetBoxedStatusCode(int statusCode)
...
+        {
+            string[] strings = LazyInitializer.EnsureInitialized(ref s_statusCodeStrings, static () => new string[512]);
```
Comment (karelz) [inline src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs:226]: Would it make sense to save size by offsetting indexes by 400?
Likely not needed, but perhaps in future? (we would have to change the method to `GetStatusCode4xx5xxString`)

**Review by karelz — APPROVED** [review]

File: src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs (hunk 1 above)
Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs:224]: Would we do it in 8.0 as well? If not I would suggest to do that in a separate PR.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs:224]: Or delete it regardless; the cache size is currently 300, so all success status code are covered.

I'm not against also raising the size, but we do need to be careful with it creeping up too high.

Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs:224]: This method will be never called with success status codes. (I will address https://github.com/dotnet/runtime/pull/91910/files#r1323214809.)

**Review by lmolkova — APPROVED** [review]
:shipit:

**Review by MihaZupan — APPROVED** [review]

**Review by karelz — APPROVED** [review]

**Review by noahfalk — APPROVED** [review]
Looks great 👍

**Comment by antonfirsov** [comment]
CI failures are #90639, #91910 + many unrelated failures in `runtime-extra-platforms`.

**Pull request merged**